File: kuber-client-csharp/examples-standalone/RabbitMqRequestResponseTest.cs
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber RabbitMQ Request/Response Test Client (v1.7.5)
 * 
 * Tests the Kuber Request/Response messaging feature via RabbitMQ AMQP.
 * 
 * Prerequisites:
 * - RabbitMQ running on localhost:5672
 * - Kuber server running with messaging enabled
 * - RabbitMQ.Client NuGet package:
 *     dotnet add package RabbitMQ.Client
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;
using System.Threading.Tasks;
using RabbitMQ.Client;
using RabbitMQ.Client.Events;

namespace Kuber.Client.Examples
{
    /// <summary>
    /// Test client for Kuber's RabbitMQ-based Request/Response messaging.
    /// </summary>
    public class RabbitMqRequestResponseTest
    {
        // Configuration
        private const string RabbitMqHost = "localhost";
        private const int RabbitMqPort = 5672;
        private const string RabbitMqUser = "guest";
        private const string RabbitMqPass = "guest";
        private const string RabbitMqVhost = "/";
        
        private const string RequestQueue = "ccs_cache_request";
        private const string ResponseQueue = "ccs_cache_response";
        
        // Your Kuber API key
        private const string ApiKey = "kub_your_api_key_here";
        private const string TestRegion = "default";
        
        // RabbitMQ objects
        private IConnection? _connection;
        private IModel? _producerChannel;
        private IModel? _consumerChannel;
        
        // State
        private readonly ConcurrentDictionary<string, TaskCompletionSource<JsonNode>> _pendingRequests = new();
        private readonly List<JsonNode> _receivedResponses = new();
        private readonly object _responseLock = new();
        
        public static async Task Main(string[] args)
        {
            Console.WriteLine("╔════════════════════════════════════════════════════════════════════╗");
            Console.WriteLine("║  Kuber RabbitMQ Request/Response Test Client (C#)                  ║");
            Console.WriteLine("║  Tests cache operations via RabbitMQ AMQP                          ║");
            Console.WriteLine("╚════════════════════════════════════════════════════════════════════╝");
            Console.WriteLine();
            
            var client = new RabbitMqRequestResponseTest();
            try
            {
                await client.RunAsync();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error: {ex.Message}");
                Console.Error.WriteLine(ex.StackTrace);
            }
        }
        
        public async Task RunAsync()
        {
            Console.WriteLine($"RabbitMQ Host: {RabbitMqHost}:{RabbitMqPort}");
            Console.WriteLine($"Virtual Host: {RabbitMqVhost}");
            Console.WriteLine($"Request Queue: {RequestQueue}");
            Console.WriteLine($"Response Queue: {ResponseQueue}");
            Console.WriteLine();
            
            try
            {
                Connect();
                StartResponseConsumer();
                
                await Task.Delay(1000);
                
                await RunTestSuiteAsync();
                
                Console.WriteLine("\nWaiting for responses (15 seconds)...");
                await Task.Delay(15000);
            }
            finally
            {
                Shutdown();
                PrintSummary();
            }
            
            Console.WriteLine("\nTest complete.");
        }
        
        private void Connect()
        {
            Console.WriteLine("Connecting to RabbitMQ...");
            
            var factory = new ConnectionFactory
            {
                HostName = RabbitMqHost,
                Port = RabbitMqPort,
                UserName = RabbitMqUser,
                Password = RabbitMqPass,
                VirtualHost = RabbitMqVhost
            };
            
            _connection = factory.CreateConnection();
            _producerChannel = _connection.CreateModel();
            _consumerChannel = _connection.CreateModel();
            
            // Declare queues
            _producerChannel.QueueDeclare(RequestQueue, durable: true, exclusive: false, autoDelete: false);
            _consumerChannel.QueueDeclare(ResponseQueue, durable: true, exclusive: false, autoDelete: false);
            
            Console.WriteLine("✓ Connected to RabbitMQ");
        }
        
        private void StartResponseConsumer()
        {
            _consumerChannel!.BasicQos(0, 10, false);
            
            var consumer = new EventingBasicConsumer(_consumerChannel);
            consumer.Received += (model, ea) =>
            {
                try
                {
                    var body = Encoding.UTF8.GetString(ea.Body.ToArray());
                    var response = JsonNode.Parse(body);
                    if (response == null) return;
                    
                    // Parse nested response
                    var requestNode = response["request"];
                    var responseNode = response["response"];
                    
                    var messageId = requestNode?["message_id"]?.GetValue<string>() ?? "unknown";
                    var success = responseNode?["success"]?.GetValue<bool>() ?? false;
                    var operation = requestNode?["operation"]?.GetValue<string>() ?? "unknown";
                    var processingTime = response["processing_time_ms"]?.GetValue<long>() ?? 0;
                    
                    Console.WriteLine();
                    Console.WriteLine("════════════════════════════════════════════════════════════");
                    Console.WriteLine("📥 RESPONSE RECEIVED");
                    Console.WriteLine($"   Message ID: {messageId}");
                    Console.WriteLine($"   Success: {success}");
                    Console.WriteLine($"   Operation: {operation}");
                    Console.WriteLine($"   Processing Time: {processingTime}ms");
                    
                    if (success)
                    {
                        var result = responseNode?["result"]?.ToJsonString();
                        if (!string.IsNullOrEmpty(result))
                        {
                            Console.WriteLine(result.Length > 100 
                                ? $"   Result: {result[..100]}..." 
                                : $"   Result: {result}");
                        }
                    }
                    else
                    {
                        var errorCode = responseNode?["error_code"]?.GetValue<string>() ?? "UNKNOWN";
                        var error = responseNode?["error"]?.GetValue<string>() ?? "Unknown error";
                        Console.WriteLine($"   Error Code: {errorCode}");
                        Console.WriteLine($"   Error: {error}");
                    }
                    Console.WriteLine("════════════════════════════════════════════════════════════");
                    
                    lock (_responseLock)
                    {
                        _receivedResponses.Add(response);
                    }
                    
                    if (_pendingRequests.TryRemove(messageId, out var tcs))
                    {
                        tcs.SetResult(response);
                    }
                    
                    _consumerChannel.BasicAck(ea.DeliveryTag, false);
                }
                catch (Exception ex)
                {
                    Console.Error.WriteLine($"Error processing message: {ex.Message}");
                    _consumerChannel.BasicNack(ea.DeliveryTag, false, false);
                }
            };
            
            _consumerChannel.BasicConsume(ResponseQueue, false, consumer);
            Console.WriteLine("✓ Response consumer started");
        }
        
        private async Task RunTestSuiteAsync()
        {
            Console.WriteLine("\n════════════════════════════════════════════════════════════");
            Console.WriteLine("STARTING TEST SUITE");
            Console.WriteLine("════════════════════════════════════════════════════════════");
            
            // Test 1: SET
            Console.WriteLine("\n--- Test 1: SET operation ---");
            var setRequest = CreateRequest("SET", TestRegion, "rabbitmq-test:user:1001");
            setRequest["value"] = new JsonObject
            {
                ["name"] = "John Doe",
                ["email"] = "john@example.com",
                ["broker"] = "rabbitmq"
            };
            setRequest["ttl"] = 300;
            PublishRequest(setRequest);
            await Task.Delay(1000);
            
            // Test 2: GET
            Console.WriteLine("\n--- Test 2: GET operation ---");
            PublishRequest(CreateRequest("GET", TestRegion, "rabbitmq-test:user:1001"));
            await Task.Delay(1000);
            
            // Test 3: EXISTS
            Console.WriteLine("\n--- Test 3: EXISTS operation ---");
            PublishRequest(CreateRequest("EXISTS", TestRegion, "rabbitmq-test:user:1001"));
            await Task.Delay(1000);
            
            // Test 4: SET another
            Console.WriteLine("\n--- Test 4: SET another key ---");
            var setRequest2 = CreateRequest("SET", TestRegion, "rabbitmq-test:user:1002");
            setRequest2["value"] = new JsonObject
            {
                ["name"] = "Jane Smith",
                ["role"] = "admin"
            };
            PublishRequest(setRequest2);
            await Task.Delay(1000);
            
            // Test 5: KEYS
            Console.WriteLine("\n--- Test 5: KEYS operation ---");
            PublishRequest(CreateRequest("KEYS", TestRegion, "rabbitmq-test:*"));
            await Task.Delay(1000);
            
            // Test 6: DELETE
            Console.WriteLine("\n--- Test 6: DELETE operation ---");
            PublishRequest(CreateRequest("DELETE", TestRegion, "rabbitmq-test:user:1002"));
            await Task.Delay(1000);
            
            // Test 7: GET non-existent
            Console.WriteLine("\n--- Test 7: GET non-existent key ---");
            PublishRequest(CreateRequest("GET", TestRegion, "rabbitmq-test:nonexistent:key"));
            await Task.Delay(1000);
            
            Console.WriteLine("\n✓ All requests sent");
        }
        
        private JsonObject CreateRequest(string operation, string region, string key)
        {
            var messageId = $"req-{Guid.NewGuid().ToString()[..12]}";
            
            var request = new JsonObject
            {
                ["operation"] = operation,
                ["region"] = region,
                ["key"] = key,
                ["api_key"] = ApiKey,
                ["message_id"] = messageId
            };
            
            _pendingRequests[messageId] = new TaskCompletionSource<JsonNode>();
            return request;
        }
        
        private void PublishRequest(JsonObject request)
        {
            var messageId = request["message_id"]!.GetValue<string>();
            var requestJson = request.ToJsonString();
            
            Console.WriteLine("📤 SENDING REQUEST");
            Console.WriteLine($"   Message ID: {messageId}");
            Console.WriteLine($"   Operation: {request["operation"]}");
            Console.WriteLine($"   Key: {request["key"]}");
            
            var props = _producerChannel!.CreateBasicProperties();
            props.ContentType = "application/json";
            props.DeliveryMode = 2;
            props.MessageId = messageId;
            
            var body = Encoding.UTF8.GetBytes(requestJson);
            _producerChannel.BasicPublish("", RequestQueue, props, body);
            
            Console.WriteLine("   ✓ Sent");
        }
        
        private void Shutdown()
        {
            Console.WriteLine("\nShutting down...");
            
            try
            {
                _producerChannel?.Close();
                _consumerChannel?.Close();
                _connection?.Close();
            }
            catch (Exception ex)
            {
                Console.Error.WriteLine($"Error closing connections: {ex.Message}");
            }
        }
        
        private void PrintSummary()
        {
            Console.WriteLine("\n════════════════════════════════════════════════════════════");
            Console.WriteLine("TEST SUMMARY");
            Console.WriteLine("════════════════════════════════════════════════════════════");
            
            lock (_responseLock)
            {
                Console.WriteLine($"Total responses received: {_receivedResponses.Count}");
            }
            Console.WriteLine($"Pending requests (no response): {_pendingRequests.Count}");
            
            if (!_pendingRequests.IsEmpty)
            {
                Console.WriteLine("\nRequests without responses:");
                foreach (var messageId in _pendingRequests.Keys)
                {
                    Console.WriteLine($"  - {messageId}");
                }
            }
            
            int successCount, errorCount;
            lock (_responseLock)
            {
                successCount = _receivedResponses.Count(r =>
                {
                    var responseNode = r["response"];
                    return responseNode?["success"]?.GetValue<bool>() ?? false;
                });
                errorCount = _receivedResponses.Count - successCount;
            }
            
            Console.WriteLine($"\nSuccessful operations: {successCount}");
            Console.WriteLine($"Failed operations: {errorCount}");
            Console.WriteLine("════════════════════════════════════════════════════════════");
        }
    }
}

File: kuber-client-java/examples/RabbitMqRequestResponseTest.java
/*
 * Copyright © 2025-2030, All Rights Reserved
 * Ashutosh Sinha | Email: ajsinha@gmail.com
 *
 * Kuber RabbitMQ Request/Response Test Client (v1.7.9)
 * 
 * Tests the Kuber Request/Response messaging feature via RabbitMQ AMQP.
 * Publishes cache operation requests and subscribes to responses.
 *
 * Prerequisites:
 * - RabbitMQ running on localhost:5672
 * - Kuber server running with messaging enabled
 * - amqp-client dependency in pom.xml:
 *     <dependency>
 *         <groupId>com.rabbitmq</groupId>
 *         <artifactId>amqp-client</artifactId>
 *         <version>5.20.0</version>
 *     </dependency>
 */
package com.kuber.client.examples;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.rabbitmq.client.*;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Test client for Kuber's RabbitMQ-based Request/Response messaging.
 * 
 * @version 1.8.1
 */
public class RabbitMqRequestResponseTest {
    
    // Configuration
    private static final String RABBITMQ_HOST = "localhost";
    private static final int RABBITMQ_PORT = 5672;
    private static final String RABBITMQ_USER = "guest";
    private static final String RABBITMQ_PASS = "guest";
    private static final String RABBITMQ_VHOST = "/";
    
    private static final String REQUEST_QUEUE = "ccs_cache_request";
    private static final String RESPONSE_QUEUE = "ccs_cache_response";
    
    // Your Kuber API key
    private static final String API_KEY = "kub_your_api_key_here";
    private static final String TEST_REGION = "default";
    
    // JSON mapper
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    // RabbitMQ objects
    private Connection connection;
    private Channel producerChannel;
    private Channel consumerChannel;
    
    // State
    private final AtomicBoolean running = new AtomicBoolean(true);
    private final Map<String, CompletableFuture<JsonNode>> pendingRequests = new ConcurrentHashMap<>();
    private final List<JsonNode> receivedResponses = new CopyOnWriteArrayList<>();
    
    public static void main(String[] args) {
        System.out.println("╔════════════════════════════════════════════════════════════════════╗");
        System.out.println("║  Kuber RabbitMQ Request/Response Test Client (Java)                ║");
        System.out.println("║  Tests cache operations via RabbitMQ AMQP                          ║");
        System.out.println("╚════════════════════════════════════════════════════════════════════╝");
        System.out.println();
        
        RabbitMqRequestResponseTest client = new RabbitMqRequestResponseTest();
        try {
            client.run();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
    
    public void run() throws Exception {
        System.out.println("RabbitMQ Host: " + RABBITMQ_HOST + ":" + RABBITMQ_PORT);
        System.out.println("Virtual Host: " + RABBITMQ_VHOST);
        System.out.println("Request Queue: " + REQUEST_QUEUE);
        System.out.println("Response Queue: " + RESPONSE_QUEUE);
        System.out.println();
        
        try {
            connect();
            startResponseConsumer();
            
            Thread.sleep(1000); // Wait for consumer setup
            
            runTestSuite();
            
            System.out.println("\nWaiting for responses (15 seconds)...");
            Thread.sleep(15000);
            
        } finally {
            shutdown();
            printSummary();
        }
        
        System.out.println("\nTest complete.");
    }
    
    private void connect() throws Exception {
        System.out.println("Connecting to RabbitMQ...");
        
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(RABBITMQ_HOST);
        factory.setPort(RABBITMQ_PORT);
        factory.setUsername(RABBITMQ_USER);
        factory.setPassword(RABBITMQ_PASS);
        factory.setVirtualHost(RABBITMQ_VHOST);
        
        connection = factory.newConnection();
        producerChannel = connection.createChannel();
        consumerChannel = connection.createChannel();
        
        // Declare queues
        producerChannel.queueDeclare(REQUEST_QUEUE, true, false, false, null);
        consumerChannel.queueDeclare(RESPONSE_QUEUE, true, false, false, null);
        
        System.out.println("✓ Connected to RabbitMQ");
    }
    
    private void startResponseConsumer() throws Exception {
        consumerChannel.basicQos(10);
        
        DeliverCallback deliverCallback = (consumerTag, delivery) -> {
            try {
                String body = new String(delivery.getBody(), StandardCharsets.UTF_8);
                JsonNode response = objectMapper.readTree(body);
                
                // Parse nested response
                JsonNode requestNode = response.get("request");
                JsonNode responseNode = response.get("response");
                
                String messageId = requestNode != null && requestNode.has("message_id") ?
                    requestNode.get("message_id").asText() : "unknown";
                boolean success = responseNode != null && responseNode.has("success") &&
                    responseNode.get("success").asBoolean();
                String operation = requestNode != null && requestNode.has("operation") ?
                    requestNode.get("operation").asText() : "unknown";
                long processingTime = response.has("processing_time_ms") ?
                    response.get("processing_time_ms").asLong() : 0;
                
                System.out.println("\n════════════════════════════════════════════════════════════");
                System.out.println("📥 RESPONSE RECEIVED");
                System.out.println("   Message ID: " + messageId);
                System.out.println("   Success: " + success);
                System.out.println("   Operation: " + operation);
                System.out.println("   Processing Time: " + processingTime + "ms");
                
                if (success) {
                    if (responseNode != null && responseNode.has("result")) {
                        String result = responseNode.get("result").toString();
                        if (result.length() > 100) {
                            System.out.println("   Result: " + result.substring(0, 100) + "...");
                        } else {
                            System.out.println("   Result: " + result);
                        }
                    }
                } else {
                    if (responseNode != null) {
                        String errorCode = responseNode.has("error_code") ?
                            responseNode.get("error_code").asText() : "UNKNOWN";
                        String error = responseNode.has("error") ?
                            responseNode.get("error").asText() : "Unknown error";
                        System.out.println("   Error Code: " + errorCode);
                        System.out.println("   Error: " + error);
                    }
                }
                System.out.println("════════════════════════════════════════════════════════════");
                
                receivedResponses.add(response);
                
                CompletableFuture<JsonNode> future = pendingRequests.remove(messageId);
                if (future != null) {
                    future.complete(response);
                }
                
                consumerChannel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
                
            } catch (Exception e) {
                System.err.println("Error processing message: " + e.getMessage());
                consumerChannel.basicNack(delivery.getEnvelope().getDeliveryTag(), false, false);
            }
        };
        
        consumerChannel.basicConsume(RESPONSE_QUEUE, false, deliverCallback, consumerTag -> {});
        System.out.println("✓ Response consumer started");
    }
    
    private void runTestSuite() throws Exception {
        System.out.println("\n════════════════════════════════════════════════════════════");
        System.out.println("STARTING TEST SUITE");
        System.out.println("════════════════════════════════════════════════════════════");
        
        // Test 1: SET
        System.out.println("\n--- Test 1: SET operation ---");
        ObjectNode setRequest = createRequest("SET", TEST_REGION, "rabbitmq-test:user:1001");
        ObjectNode userData = objectMapper.createObjectNode();
        userData.put("name", "John Doe");
        userData.put("email", "john@example.com");
        userData.put("broker", "rabbitmq");
        setRequest.set("value", userData);
        setRequest.put("ttl", 300);
        publishRequest(setRequest);
        Thread.sleep(1000);
        
        // Test 2: GET
        System.out.println("\n--- Test 2: GET operation ---");
        publishRequest(createRequest("GET", TEST_REGION, "rabbitmq-test:user:1001"));
        Thread.sleep(1000);
        
        // Test 3: EXISTS
        System.out.println("\n--- Test 3: EXISTS operation ---");
        publishRequest(createRequest("EXISTS", TEST_REGION, "rabbitmq-test:user:1001"));
        Thread.sleep(1000);
        
        // Test 4: SET another
        System.out.println("\n--- Test 4: SET another key ---");
        ObjectNode setRequest2 = createRequest("SET", TEST_REGION, "rabbitmq-test:user:1002");
        ObjectNode userData2 = objectMapper.createObjectNode();
        userData2.put("name", "Jane Smith");
        userData2.put("role", "admin");
        setRequest2.set("value", userData2);
        publishRequest(setRequest2);
        Thread.sleep(1000);
        
        // Test 5: KEYS
        System.out.println("\n--- Test 5: KEYS operation ---");
        publishRequest(createRequest("KEYS", TEST_REGION, "rabbitmq-test:*"));
        Thread.sleep(1000);
        
        // Test 6: DELETE
        System.out.println("\n--- Test 6: DELETE operation ---");
        publishRequest(createRequest("DELETE", TEST_REGION, "rabbitmq-test:user:1002"));
        Thread.sleep(1000);
        
        // Test 7: GET non-existent
        System.out.println("\n--- Test 7: GET non-existent key ---");
        publishRequest(createRequest("GET", TEST_REGION, "rabbitmq-test:nonexistent:key"));
        Thread.sleep(1000);
        
        System.out.println("\n✓ All requests sent");
    }
    
    private ObjectNode createRequest(String operation, String region, String key) {
        String messageId = "req-" + UUID.randomUUID().toString().substring(0, 12);
        
        ObjectNode request = objectMapper.createObjectNode();
        request.put("operation", operation);
        request.put("region", region);
        request.put("key", key);
        request.put("api_key", API_KEY);
        request.put("message_id", messageId);
        
        pendingRequests.put(messageId, new CompletableFuture<>());
        
        return request;
    }
    
    private void publishRequest(ObjectNode request) throws Exception {
        String messageId = request.get("message_id").asText();
        String requestJson = objectMapper.writeValueAsString(request);
        
        System.out.println("📤 SENDING REQUEST");
        System.out.println("   Message ID: " + messageId);
        System.out.println("   Operation: " + request.get("operation").asText());
        System.out.println("   Key: " + request.get("key").asText());
        
        AMQP.BasicProperties props = new AMQP.BasicProperties.Builder()
            .contentType("application/json")
            .deliveryMode(2)
            .messageId(messageId)
            .build();
        
        producerChannel.basicPublish("", REQUEST_QUEUE, props, requestJson.getBytes(StandardCharsets.UTF_8));
        System.out.println("   ✓ Sent");
    }
    
    private void shutdown() {
        System.out.println("\nShutting down...");
        running.set(false);
        
        try {
            if (producerChannel != null) producerChannel.close();
            if (consumerChannel != null) consumerChannel.close();
            if (connection != null) connection.close();
        } catch (Exception e) {
            System.err.println("Error closing connections: " + e.getMessage());
        }
    }
    
    private void printSummary() {
        System.out.println("\n════════════════════════════════════════════════════════════");
        System.out.println("TEST SUMMARY");
        System.out.println("════════════════════════════════════════════════════════════");
        System.out.println("Total responses received: " + receivedResponses.size());
        System.out.println("Pending requests (no response): " + pendingRequests.size());
        
        if (!pendingRequests.isEmpty()) {
            System.out.println("\nRequests without responses:");
            for (String messageId : pendingRequests.keySet()) {
                System.out.println("  - " + messageId);
            }
        }
        
        long successCount = receivedResponses.stream()
            .filter(r -> {
                JsonNode responseNode = r.get("response");
                return responseNode != null && responseNode.has("success") && responseNode.get("success").asBoolean();
            })
            .count();
        long errorCount = receivedResponses.size() - successCount;
        
        System.out.println("\nSuccessful operations: " + successCount);
        System.out.println("Failed operations: " + errorCount);
        System.out.println("════════════════════════════════════════════════════════════");
    }
}
